tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,265 @@
+};
+use tracing_core::{
+    dispatcher,
+    field::FieldSet,
+    span::{self, Current, Id},
+    Event, Interest, Metadata, Subscriber,
+};
+
+#[derive(Debug)]
+pub struct Registry {
+    spans: Slab<Data>,
+}
+
+thread_local! {
+    static CONTEXT: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+#[derive(Debug)]
+pub struct Data {
+    metadata: &'static Metadata<'static>,
+    parent: Option<Id>,
+    children: Vec<Id>,
+    ref_count: AtomicUsize,
+    extensions: RwLock<Extensions>,
+}
+
+// === impl Registry ===
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self { spans: Slab::new() }
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: Data) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, Data>> {
+        self.spans.get(id_to_idx(id))
+    }
+}
+
+thread_local! {
+    static CURRENT_SPANS: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let parent = attrs.parent().map(|id| self.clone_span(id));
+        let s = Data {
+            metadata: attrs.metadata(),
+            parent,
+            children: vec![],
+            ref_count: AtomicUsize::new(1),
+            extensions: RwLock::new(Extensions::new()),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    #[inline]
+    fn record(&self, _: &span::Id, _: &span::Record<'_>) {
+        // TODO(david)—implement this.
+    }
+
+    fn record_follows_from(&self, _: &span::Id, _: &span::Id) {
+        // TODO(david)—implement this.
+    }
+
+    fn event(&self, _: &Event<'_>) {
+        // TODO(david)—implement this.
+    }
+
+    fn enter(&self, id: &span::Id) {
+        CURRENT_SPANS.with(|spans| {
+            spans.borrow_mut().push(self.clone_span(id));
+        })
+    }
+
+    fn exit(&self, id: &span::Id) {
+        if let Some(id) = CURRENT_SPANS.with(|spans| spans.borrow_mut().pop(id)) {
+            dispatcher::get_default(|dispatch| dispatch.try_close(id.clone()));
+        }
+    }
+
+    fn clone_span(&self, id: &span::Id) -> span::Id {
+        let refs = self
+            .get(&id)
+            .unwrap_or_else(|| panic!("tried to clone {:?}, but no span exists with that ID", id))
+            .ref_count
+            .fetch_add(1, Ordering::Relaxed);
+        assert!(refs != 0, "tried to clone a span that already closed");
+        id.clone()
+    }
+
+    fn current_span(&self) -> Current {
+        CURRENT_SPANS.with(|spans| {
```

Comment:
This looks correct, but I think it could be a lot simpler:
```rust
CURRENT_SPANS.with(|spans| {
    let spans = spans.borrow();
    let id = spans.current()?;
    let span = self.get(current)?;
    Some(Current::new(id.clone(), span.metadata))
}).unwrap_or_else(Current::none)
```

Reply:
done, added.